File: backend/src/main/java/com/sakuya/backend/notification/AppNotification.java
package com.sakuya.backend.notification;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.UUID;

/**
 * AppNotification.java
 * 职责说明：保存用户可追溯的一条业务通知，是通知中心与未读数的唯一真相源。
 * 执行流程：业务事务创建通知 -> Outbox 异步派发 FCM -> 客户端通过 REST 同步并标记已读。
 */
@Entity
@Table(name = "app_notifications", uniqueConstraints = @UniqueConstraint(name = "uk_notification_source", columnNames = "source_key"),
    indexes = {@Index(name = "idx_notification_recipient_created", columnList = "recipient_id,created_at"), @Index(name = "idx_notification_recipient_read", columnList = "recipient_id,read_at")})
public class AppNotification {
    public enum Type { FEED_COMMENT, FEED_REPLY, FEED_LIKE, FEED_FAVORITE, NEW_FOLLOWER, FRIEND_REQUEST, FRIEND_ACCEPTED, CHAPTER_UPDATE }
    public enum TargetType { FEED_POST, USER, FRIEND_REQUEST, BOOK_CHAPTER }

    @Id private UUID id;
    @Column(name = "recipient_id", nullable = false) private UUID recipientId;
    @Column(name = "actor_id", nullable = false) private UUID actorId;
    @Enumerated(EnumType.STRING) @Column(nullable = false, length = 32) private Type type;
    @Column(nullable = false, length = 120) private String title;
    @Column(nullable = false, length = 500) private String content;
    @Enumerated(EnumType.STRING) @Column(name = "target_type", nullable = false, length = 24) private TargetType targetType;
    @Column(name = "target_id", nullable = false) private String targetId;
    @Column(name = "target_user_id") private UUID targetUserId;
    @Column(name = "aggregate_key", length = 160) private String aggregateKey;
    @Column(name = "source_key", nullable = false, length = 220) private String sourceKey;
    @Column(name = "created_at", nullable = false, updatable = false) private Instant createdAt;
    @Column(name = "read_at") private Instant readAt;

    protected AppNotification() {}
    public AppNotification(UUID recipientId, UUID actorId, Type type, String title, String content, TargetType targetType,
                           String targetId, UUID targetUserId, String aggregateKey, String sourceKey) {
        this.id = UUID.randomUUID(); this.recipientId = recipientId; this.actorId = actorId; this.type = type;
        this.title = title; this.content = content; this.targetType = targetType; this.targetId = targetId;
        this.targetUserId = targetUserId; this.aggregateKey = aggregateKey; this.sourceKey = sourceKey; this.createdAt = Instant.now();
    }
    public UUID getId(){return id;} public UUID getRecipientId(){return recipientId;} public UUID getActorId(){return actorId;}
    public Type getType(){return type;} public String getTitle(){return title;} public String getContent(){return content;}
    public TargetType getTargetType(){return targetType;} public String getTargetId(){return targetId;}
    public UUID getTargetUserId(){return targetUserId;} public String getAggregateKey(){return aggregateKey;}
    public Instant getCreatedAt(){return createdAt;} public Instant getReadAt(){return readAt;}
    public void markRead(){if(readAt==null)readAt=Instant.now();}
}

File: core/model/src/main/java/com/sakuya/model/notification/AppNotification.kt
package com.sakuya.model.notification

/**
 * AppNotification.kt
 * 职责说明：定义通知中心、系统推送与导航共同识别的受控通知类型。
 * 安全约束：未知服务端类型映射为 null 并只允许打开通知中心，不把任意字符串当作应用路由。
 */
enum class NotificationType { FEED_COMMENT, FEED_REPLY, FEED_LIKE, FEED_FAVORITE, NEW_FOLLOWER, FRIEND_REQUEST, FRIEND_ACCEPTED, CHAPTER_UPDATE }
enum class NotificationTargetType { FEED_POST, USER, FRIEND_REQUEST, BOOK_CHAPTER }

data class AppNotification(
    val id: String,
    val actorId: String,
    val type: NotificationType,
    val title: String,
    val content: String,
    val targetType: NotificationTargetType,
    val targetId: String,
    val targetUserId: String?,
    val createdAt: String,
    val readAt: String?,
)

data class NotificationPreferences(
    val pushEnabled: Boolean = true,
    val feedEnabled: Boolean = true,
    val friendEnabled: Boolean = true,
)

/** 客户端只允许按固定分类清除通知，避免 UI 传入任意服务端类型集合。 */
enum class NotificationUnreadCategory { FEED_INTERACTION, NEW_FOLLOWER, FRIEND_RELATION }

data class NotificationUnreadSummary(
    val total: Int = 0,
    val feedInteraction: Int = 0,
    val newFollowers: Int = 0,
    val friendRelation: Int = 0,
)

data class FollowingUnseenSummary(val postCount: Int = 0,val authorCount: Int = 0)

/**
 * UpdateBadgeState.kt
 * 职责说明：集中表达各一级入口是否显示无数字小红点，UI 不自行拼接业务类型。
 * 组合规则：消息=聊天或任一通知；我的=动态互动、新粉丝或关注用户新动态。
 */
data class UpdateBadgeState(
    val notification: NotificationUnreadSummary = NotificationUnreadSummary(),
    val following: FollowingUnseenSummary = FollowingUnseenSummary(),
    val conversationUnreadCount: Int = 0,
) {
    val showMessage: Boolean get() = conversationUnreadCount > 0 || notification.total > 0
    val showProfile: Boolean get() = notification.feedInteraction > 0 || notification.newFollowers > 0 || following.postCount > 0
    val showProfileDynamic: Boolean get() = notification.feedInteraction > 0
    val showFollowers: Boolean get() = notification.newFollowers > 0
    val showFollowing: Boolean get() = following.postCount > 0
    val showFriendRelation: Boolean get() = notification.friendRelation > 0
}
